### PR TITLE
Hughsie/no blobs

### PR DESCRIPTION
**Pull request by hughsie**
Type of pull request:

- [ ] New plugin (Please include [new plugin checklist](https://github.com/fwupd/fwupd/wiki/New-plugin-checklist))
- [ ] Code fix
- [ ] Feature
- [ ] Documentation

### CODE REVIEW

**Comment by superm1** [comment]
Can we at least move them to fwupd test images repo? Then they're only used in CI.

**Comment by hughsie** [comment]
> Can we at least move them to fwupd test images repo

The problem is that then they bitrot. I've spent a few hours already fixing up the various things that have broken over the last 3 years just so we can re-create the binary files. If we *construct* them at build time then we notice any breakage in the commit that broke them.

**Comment by superm1** [comment]
> > Can we at least move them to fwupd test images repo
> 
> The problem is that then they bitrot. I've spent a few hours already fixing up the various things that have broken over the last 3 years just so we can re-create the binary files. If we _construct_ them at build time then we notice any breakage in the commit that broke them.

Well as they're from CI we _should_ notice commits that break them, but yeah I get your point.  I'm just saying at least for the EDK2 binaries case _maybe_ it's better.

BTW - I'd like to think that the way our build system, tarball generation and tree organization is done we wouldn't have been able to be hit by such an attack.

File: libfwupdplugin/fu-fuzzer-main.c [hunk 1]
```diff
@@ -10,13 +10,36 @@ __attribute__((weak)) extern int
...
 int
 main(int argc, char **argv)
 {
 	g_assert_nonnull(LLVMFuzzerTestOneInput);
 	if (LLVMFuzzerInitialize != NULL)
 		LLVMFuzzerInitialize(&argc, &argv);
+
+	/* do not use g_option_context_parse() here for speed */
+	if (argc == 3 && g_str_has_suffix(argv[1], ".builder.xml") &&
+	    g_str_has_suffix(argv[2], ".bin")) {
```
Comment (superm1) [inline libfwupdplugin/fu-fuzzer-main.c:25]: this feels a bit fragile - could it be done by an environment variable instead?

File: libfwupdplugin/fu-fuzzer-main.c [hunk 2]
```diff
@@ -10,13 +10,36 @@ __attribute__((weak)) extern int
...
+		g_autoptr(GBytes) blob_dst = NULL;
+		g_autoptr(GError) error = NULL;
+		blob_dst = BuildOneOutput(argv[1], &error);
+		if (blob_dst == NULL) {
+			g_printerr("Failed to build output: %s\n", error->message);
+			return 1;
```
Comment (superm1) [inline libfwupdplugin/fu-fuzzer-main.c:31]: `EXIT_FAILURE`?

File: libfwupdplugin/fu-fuzzer-main.c [hunk 3]
```diff
@@ -10,13 +10,36 @@ __attribute__((weak)) extern int
...
+		}
+		if (!g_file_set_contents(argv[2],
+					 g_bytes_get_data(blob_dst, NULL),
+					 g_bytes_get_size(blob_dst),
+					 &error)) {
+			g_printerr("Failed to save: %s\n", error->message);
+			return 1;
+		}
+		return 0;
```
Comment (superm1) [inline libfwupdplugin/fu-fuzzer-main.c:40]: `EXIT_SUCCESS`?

**Review by hughsie — COMMENTED** [review]
actually, maybe i'm being in an idiot -- iirc oss-fuzz loads the binary and then uses LLVMFuzzerTestOneInput as a weak symbol  -- maybe I can just use g_option_context_parse() without fear. maybe something like `build --xml foo --bin bar`

File: libfwupdplugin/fu-fuzzer-firmware.c.in [hunk 4]
```diff
@@ -33,3 +33,12 @@ LLVMFuzzerTestOneInput(const guint8 *data, gsize size)
 	}
 	return 0;
 }
+
+GBytes *
+BuildOneOutput(const gchar *filename, GError **error)
+{
+	g_autoptr(FuFirmware) firmware = FU_FIRMWARE(@FIRMWARENEW@);
+	if (!fu_firmware_build_from_filename(firmware, filename, error))
```
[the remark this answers is not in the document]
Comment (hughsie) [reply libfwupdplugin/fu-fuzzer-firmware.c.in:41]: Yup, I guess. Fixed!

**Comment by hughsie** [comment]
> iirc oss-fuzz loads the binary and then uses LLVMFuzzerTestOneInput as a weak symbol

I was right the first time -- the binary is re-exec'd and the GOptionEntry does slow down using afl.

**Comment by hughsie** [comment]
> plugins/thunderbolt/tests/minimal-fw.bin
> plugins/thunderbolt/tests/minimal-fw-controller.bin

These are gone now too.

**Comment by hughsie** [comment]
@superm1 this should also get us a near perfect score for *Binary-Artifacts* in https://securityscorecards.dev/viewer/?uri=github.com/fwupd/fwupd

**Review by superm1 — APPROVED** [review]

**Comment by superm1** [comment]
It looks to me that CI isn't working.
```
+ python3 fwupd/contrib/ci/oss-fuzz.py
Traceback (most recent call last):
  File "fwupd/contrib/ci/oss-fuzz.py", line 19, in <module>
    class Builder:
  File "fwupd/contrib/ci/oss-fuzz.py", line 193, in Builder
    def mkfuzztargets(self, exe: str, globstr: str) -> list[str]:
TypeError: 'type' object is not subscriptable
2024-04-08 06:50:09,474 - root - ERROR - Building fuzzers failed.
2024-04-08 06:50:09,475 - root - ERROR - Error building fuzzers for (commit: 8ed726d219a8064db2cbb12ba984f0aafe8cc8d8, pr_ref: None).
```

**Comment by superm1** [comment]
> It looks to me that CI isn't working.
> 
> ```
> + python3 fwupd/contrib/ci/oss-fuzz.py
> Traceback (most recent call last):
>   File "fwupd/contrib/ci/oss-fuzz.py", line 19, in <module>
>     class Builder:
>   File "fwupd/contrib/ci/oss-fuzz.py", line 193, in Builder
>     def mkfuzztargets(self, exe: str, globstr: str) -> list[str]:
> TypeError: 'type' object is not subscriptable
> 2024-04-08 06:50:09,474 - root - ERROR - Building fuzzers failed.
> 2024-04-08 06:50:09,475 - root - ERROR - Error building fuzzers for (commit: 8ed726d219a8064db2cbb12ba984f0aafe8cc8d8, pr_ref: None).
> ```

Maybe we should build fuzzers as part of PR CI (but not run them as part of PR CI)?

**Comment by hughsie** [comment]
> Maybe we should build fuzzers as part of PR CI (but not run them as part of PR CI)?

Ahh, that wouldn't of helped here; I accidentally used a feature only available in newer Pythons -- and I'm guessing osfuzz is using an older one. Fixed hopefully in main.

**Pull request merged**